libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthCoordinator.m

```diff
@@ -557,15 +558,65 @@ - (void)beginUserAgentFlow {
         [approvalUrl appendString:scopeString];
     }
     
+    // JWT Flow
+    if (self.credentials.jwt && self.credentials.instanceUrl) {
+        [self swapJWTWithcompletionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
+            bool swapOK = NO;
+            if (!error) {
+                NSError *jsonError = nil;
+                id json = nil;
+                
+                json = [NSJSONSerialization JSONObjectWithData:data options:0 error:&jsonError];
+                if (nil == jsonError && [json isKindOfClass:[NSDictionary class]]) {
+                    NSDictionary *dict = (NSDictionary *)json;
+                    if (dict[kSFOAuthAccessToken]) {
+                        NSString *escapedString = [approvalUrl stringByURLEncoding];
+                        NSString* approvalUrl = [NSString stringWithFormat:@"%@://%@/secur/frontdoor.jsp?sid=%@&retURL=%@", self.credentials.protocol, self.credentials.instanceUrl, dict[kSFOAuthAccessToken],escapedString];
+                        [self doLoadURL:approvalUrl withCookie:YES];
+                        swapOK = YES;
+                        self.credentials.jwt = nil;
+                    }
+                }
+            }
+            if (!swapOK) {
+                [self log:SFLogLevelInfo msg:@"Fail to swap JWT for access token"];
+                [self doLoadURL:approvalUrl withCookie:NO];
+            }
+        }];
+    }
+    else {
+        [self doLoadURL:approvalUrl withCookie:NO];
+    }
+}
+
+
+- (void)swapJWTWithcompletionHandler:(void (^)(NSData *data, NSURLResponse *response, NSError *error))completionHandler  {
+    NSString *url = [[NSString alloc] initWithFormat:@"%@://%@%@", self.credentials.protocol,
```

Comment:
@khawkins that's right, but I believe for the token endpoint, both login endpoint and instance url endpoint should work.. The reason I use instanceURL here is it's the only thing we have... For the frontdoor flow, I need the instanceURL, so in the magic link, it gives  us 2 things: JWT and instanceURL.. but think about it again, you have a valid point, I should in this case set login URL as instanceURL or maybe ask the magicLink provide login URL also, so here it should login URL, I will send a fix soon..